async-nats/src/jetstream/kv/mod.rs

```diff
@@ -0,0 +1,499 @@
+// Copyright 2020-2022 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub mod bucket;
+
+use std::{
+    collections::{self, HashSet},
+    io,
+    task::Poll,
+};
+
+use crate::HeaderValue;
+use bytes::Bytes;
+use futures::{StreamExt, TryStreamExt};
+use lazy_static::lazy_static;
+use regex::Regex;
+use time::OffsetDateTime;
+
+use crate::{header, jetstream::response, Error, Message};
+
+use self::bucket::Status;
+
+use super::{
+    consumer::DeliverPolicy,
+    stream::{RawMessage, StorageType, Stream},
+};
+
+// Helper to extract key value operation from message headers
+fn kv_operation_from_maybe_headers(maybe_headers: Option<&String>) -> Operation {
+    if let Some(headers) = maybe_headers {
+        return match headers.as_str() {
+            KV_OPERATION_DELETE => Operation::Delete,
+            KV_OPERATION_PURGE => Operation::Purge,
+            _ => Operation::Put,
+        };
+    }
+
+    Operation::Put
+}
+
+fn kv_operation_from_stream_message(message: &RawMessage) -> Operation {
+    kv_operation_from_maybe_headers(message.headers.as_ref())
+}
+lazy_static! {
+    static ref VALID_BUCKET_RE: Regex = Regex::new(r#"\A[a-zA-Z0-9_-]+\z"#).unwrap();
+    static ref VALID_KEY_RE: Regex = Regex::new(r#"\A[-/_=\.a-zA-Z0-9]+\z"#).unwrap();
+}
+
+pub(crate) const MAX_HISTORY: i64 = 64;
+const ALL_KEYS: &str = ">";
+
+const KV_OPERATION: &str = "KV-Operation";
+const KV_OPERATION_DELETE: &str = "DEL";
+const KV_OPERATION_PURGE: &str = "PURGE";
+const KV_OPERATION_PUT: &str = "PUT";
+
+const NATS_ROLLUP: &str = "Nats-Rollup";
+const ROLLUP_SUBJECT: &str = "sub";
+
+pub(crate) fn is_valid_bucket_name(bucket_name: &str) -> bool {
+    VALID_BUCKET_RE.is_match(bucket_name)
+}
+
+pub(crate) fn is_valid_key(key: &str) -> bool {
+    if key.is_empty() || key.starts_with('.') || key.ends_with('.') {
+        return false;
+    }
+
+    VALID_KEY_RE.is_match(key)
+}
+
+/// Configuration values for key value stores.
+#[derive(Debug, Default)]
+pub struct Config {
```

Comment:
Double check which options are optional, and Option accordingly? In old nats we made too much use of embedded control values over options because that was what already existed, should try not to now imo.